src/functor.jl

```diff
@@ -1,13 +1,31 @@
+function functor end
 
-functor(T, x) = (), _ -> x
+const NoChildren = Tuple{}
+
+function makeleaf(m::Module, T)
+  @eval m begin
```

Comment:
If I'm not mistaken, we should be able to do this without the `@eval` by emitting the function definition from the macro directly.

Reply:
I need more indirection for that, my macro skills are terrible, I'm almost blindly adapting the code from `@functor`

Reply:
Mine are pretty atrocious as well, but after playing around locally this might work:
```julia
macro leaf(T)
  :($Functors.functor(::Type{<:$(esc(T))}, x) = $Functors.NoChildren(), _ -> x
end
```

Reply:
```
macro leaf(T)
  :($Functors.functor(::Type{<:$(esc(T))}, x) = $Functors.NoChildren(), _ -> x)
end
```
errors with 
```julia
julia> using Functors
[ Info: Precompiling Functors [d9f16b24-f501-4c13-a1f2-28368ffc5196]
ERROR: LoadError: syntax: invalid named tuple field name "Functors.functor(::Type{<:(escape Any)}, x)" around /Users/carlo/.julia/dev/Functors/src/functor.jl:20
```

Reply:
works